sdks/java/core/src/test/java/org/apache/beam/sdk/values/RowTest.java

```diff
@@ -297,4 +299,30 @@ public void testThrowsForIncorrectNumberOfFields() {
     thrown.expect(IllegalArgumentException.class);
     Row.withSchema(type).addValues(1, "2").build();
   }
+
+  @Test
+  public void testByteArrayEquality() {
+    byte[] a0 = new byte[] {1, 2, 3, 4};
+    byte[] b0 = new byte[] {1, 2, 3, 4};
+
+    Schema schema = Schema.of(Schema.Field.of("bytes", Schema.FieldType.BYTES));
+
+    Row a = Row.withSchema(schema).addValue(a0).build();
+    Row b = Row.withSchema(schema).addValue(b0).build();
+
+    Assert.assertEquals(a, b);
+  }
+
+  @Test
+  public void testByteBufferEquality() {
+    byte[] a0 = new byte[] {1, 2, 3, 4};
+    byte[] b0 = new byte[] {1, 2, 3, 4};
+
+    Schema schema = Schema.of(Schema.Field.of("bytes", Schema.FieldType.BYTES));
```

Comment:
Need a test of many deeper structures.

Reply:
https://issues.apache.org/jira/browse/BEAM-5869